### ISSUE TITLE
Expose parse_string in JSONDecoder

### ISSUE THREAD

**Issue by 9b78a555-0225-4922-bbb5-ac2962d8f772** [python/cpython#74178]
BPO | [29992](https://bugs.python.org/issue29992)
--- | :---
Nosy | @rhettinger, @etrepum, @ezio-melotti, @methane, @serhiy-storchaka
Files | <li>[scanner.py](https://bugs.python.org/file47374/scanner.py "Uploaded as text/plain at 2018-01-10.13:52:33 by Adrián Orive"): Modified JSON module's scanner.py</li><li>[decoder.py](https://bugs.python.org/file47375/decoder.py "Uploaded as text/plain at 2018-01-10.13:53:47 by Adrián Orive"): Modified JSON module's decoder.py</li><li>[__init__.py](https://bugs.python.org/file47376/__init__.py "Uploaded as text/plain at 2018-01-10.13:54:12 by Adrián Orive"): Modified JSON module's __init__.py</li>

<sup>*Note: these values reflect the state of the issue at the time it was migrated and might not reflect the current state.*</sup>

<details><summary>Show more details</summary><p>

GitHub fields:
```python
assignee = 'https://github.com/etrepum'
closed_at = None
created_at = <Date 2017-04-05.08:22:05.524>
labels = ['3.7', 'type-feature', 'library']
title = 'Expose parse_string in JSONDecoder'
updated_at = <Date 2022-02-17.02:41:38.992>
user = 'https://bugs.python.org/oberstet'
```

bugs.python.org fields:
```python
activity = <Date 2022-02-17.02:41:38.992>
actor = 'methane'
assignee = 'bob.ippolito'
closed = False
closed_date = None
closer = None
components = ['Library (Lib)']
creation = <Date 2017-04-05.08:22:05.524>
creator = 'oberstet'
dependencies = []
files = ['47374', '47375', '47376']
hgrepos = []
issue_num = 29992
keywords = []
message_count = 14.0
messages = ['291167', '291172', '291173', '291180', '291186', '292707', '292782', '309763', '309765', '309766', '309819', '309820', '413338', '413378']
nosy_count = 8.0
nosy_names = ['rhettinger', 'bob.ippolito', 'ezio.melotti', 'methane', 'oberstet', 'serhiy.storchaka', 'Levi Cameron', 'Adri\xc3\xa1n Orive']
pr_nums = []
priority = 'normal'
resolution = None
stage = None
status = 'open'
superseder = None
type = 'enhancement'
url = 'https://bugs.python.org/issue29992'
versions = ['Python 3.7']
```

</p></details>

**Comment by 9b78a555-0225-4922-bbb5-ac2962d8f772** [comment]
Though the JSONDecoder already has all the hooks internally to allow for a custom parse_string (https://github.com/python/cpython/blob/master/Lib/json/decoder.py#L330), this currently is not exposed in the constructor JSONDecoder.\_\_init__.

It would be nice to expose it. Currently, I need to do hack it: https://gist.github.com/oberstet/fa8b8e04b8d532912bd616d9db65101a

**Comment by serhiy-storchaka** [comment]
JSONDecoder constructor already has too much parameters. Adding new parameters will decrease usability. For such uncommon case I think overriding a method in a subclass is the best solution.

**Comment by 9b78a555-0225-4922-bbb5-ac2962d8f772** [comment]
I agree, my use case is probably exotic: transparent roundtripping of binaries over JSON using a beginning \\0 byte marker to distinguish plain string and base64 encoded binaries.

FWIW, I do think however that adding "parse_string" kw param to the ctor of JSONDecoder would at least fit the current approach: there are parse_xxx parameters for all the other things already.

If overriding string parsing would be via subclassing, while all the others stay with the kw parameter approach, that could be slightly confusing too, because it looses on consistency.

Switching everything to subclassing/overriding for \_all_ parse_XXX is I guess a no go, because it breaks existing stuff?

> For me in my situation, it'll be messy anyways, because I need to support Py2 and 3, and CPy and PyPy .. I just filed the issue for "completeness".

**Comment by rhettinger** [comment]
I agree with Serhiy that the JSON module is already complex enough that adding more features will have a negative net effect on usability.

Bob, what do you think?

**Comment by 8bb2fe13-d245-4fef-8aa2-70d7108db3c3** [comment]
I agree with that sentiment. If we were to want to support this use case I would rather put together a coherent way to augment the parsing/encoding of anything than bolt it on to what we have.

**Comment by 05f844fe-3a6a-471b-97a6-c20a8a054910** [comment]
A less exotic use case than oberstet's: Converting code from python 2 -\> 3 and trying to maintain python 2 behaviour of returning all strings as bytes rather than unicode strings.

obsertet's solution works but is very much tied to the current implementation.

**Comment by 8bb2fe13-d245-4fef-8aa2-70d7108db3c3** [comment]
That's not a very convincing argument. Python 2 only returns byte strings if the input is a byte string and the contents of the string are all ASCII. Facilitating that sort of behavior in 3 would probably cause more issues than it solves.
